### ISSUE TITLE
Interact with Vagrant VMs though REST Interface

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#979]
We would like to further streamline https://github.com/Moonshine-IDE/NSFConverterPortalProject/issues/6 by having Moonshine communicate with the Vagrant VM through a simple REST interface.   The REST interface will be a separate project, so this issue will be more focused on the interaction with Moonshine.

TODO:  writeup an example workflow

REST interface actions:
- Push file
- Pull file
- Run command
- Check status on long-term action?

Here are some starting points for making the REST interface:
- Spark:  https://sparkjava.com/
- RESTx:  http://restx.io/
- https://syntaxcorrect.com/Java/5_Ultra_Lightweight_Http_Server_Implementations_in_Java_for_Blazing_Fast_Microservices_APIs_or_Even_Websites

Branch:  features/issue_979_vagrant_REST_interaction
macOS Build:  https://moonshine-ide.com/downloads/releases/dev/macos/branches/features/issue_979_vagrant_REST_interaction
Windows Build:  https://moonshine-ide.com/downloads/releases/dev/windows/branches/features/issue_979_vagrant_REST_interaction

**Comment by JoelProminic** [comment]
I reviewed this more with @JustinProminic, and we are favoring using [Rapidoid](https://www.rapidoid.org/).  What we really want here is an application that will let us quickly build up a REST interface, so I'll see how it looks after going through some examples.   If we will need to invest a lot of time to learn a new framework, it may be better to return to a familiar framework like Grails or Primefaces/JSF, if it lets us get done faster.

**Comment by JoelProminic** [comment]
We also discussed implementing this using SSH and SCP instead of the REST interface.   We have a Java library (JSch) which we use for this, so it would be familiar and powerful.  However:
- We have had problems previously using Java applications as part of our Flex applications, I think with code signing.
- Integration with Java would probably take more work on the Actionscript side than REST
- This won't be practical in a security sense if we want to reuse the code for a production server (though this could be a problem for our actions in general).  We could replace SCP with an FTP server like CrushFTP for more security options, but this could result in a lot of differences.

So, I'd like to stick with the REST interface unless this turns out to be way harder than expected.

**Comment by JoelProminic** [comment]
Here is some more brainstorming about how we would use the REST interface:
- Build an On Disk Project in the VM (to avoid Compiling ODP hang and local dependencies)
    - upload project to VM
    - run maven command on VM
    - download the NSF or other files?
- Deploy a database to a local server for local testing
- Generate an Update Site on the server (needed for Windows, current logic has Notes client depenency on macOS)
    - Clone update site repository on VM (or have this in advance)
    - Run Maven script
    - Zip the update site
    - Download update site locally and unzip to MoonshineSDKs/Domino
- Test local agents by deploying to local server with DXL Importer
    - Upload project to VM
    - Run Ant scripts to import agent(s) to local server
- Converter databases to Domino Visual Editor projects (NSFConverterPortalProject)
    - Upload NSF to server
    - run command to convert database?  See discussion below
    - Download project from server
    - Open project in Moonshine

**Comment by JoelProminic** [comment]
We discussed this today, and we would like to implement it with Grails so that we use a familiar technology.

**Comment by JoelProminic** [comment]
@solaechea started on the REST interface here:  https://github.com/Moonshine-IDE/REST-Interface

I'll plan to create some issues on this repository.

**Comment by JoelProminic** [comment]
I have an issue here for testing this tool with NSFConverter:  https://github.com/Moonshine-IDE/NSFConverterPortalProject/issues/13

**Comment by JoelProminic** [comment]
Here is the process for interacting with the REST interface.  I use curl as an example command, but you should use Actionscript or Haxe APIs instead.  There are some missing details in the `/task` response that I need to discuss with @solaechea (https://github.com/Moonshine-IDE/REST-Interface/issues/2#issuecomment-1171632736), but this is at least an up-to-date summary of the conversion process.

First, the NSF needs to be uploaded.  The `file` parameter is the encoded file data - I couldn't export an example for this with curl or PostMan:

    # URL:  /file/upload
    # Content-Type:  multipart/form-data
    # Parameter:  file (the attachment data)
    curl -X "POST" "http://localhost:8095/file/upload" \
         -H "Content-Type: multipart/form-data" \
         -F "file=@/Users/joelanderson/Documents/vagrant-workspace/NSFConverterPortalProject/TestFieldOptions_1008_4.nsf"


The output looks like this.  Save the "path" to use in the next command (`%nsf-path%`):

    {
        "path":"/tmp/restinterface/upload/TestFieldOptions_1008_4.nsf",
        "size":524288
    }


Next, run the conversion command:

    # URL:  /task
    # Content-Type:  application/x-www-form-urlencoded
    # Parameter:  command (the command to run
    #   - "/opt/nsf-converter-portal/scripts/nsf-odp-convert.sh".  This is a hard-coded path to the script.  Vagrant will enforce this location.
    #   - "%nsf-path%" - collected above.  
    #   - "%output-file%" - the name of the resulting file.  You can hard-code this as "result.zip"

    curl -X "POST" "http://localhost:8095/task" \
         -H "Accept: application/json" \
         -H "Content-Type: application/x-www-form-urlencoded" \
         --data-urlencode "command=/bin/bash /opt/nsf-converter-portal/scripts/nsf-odp-convert.sh '%nsf-path%' '%output-file%'"

The output looks like this.  You should collect:
- "id" - this is an ID used to check the status later (`%taskid%`)
- "workingDir" - this will be combined with `%output-file%` to get the result file later (`%working-dir%`)

```
    {
        "id":1,
        "startDate":"2022-06-27T22:31:13Z",
        "workingDir":"/tmp/restinterface/working",
        "command":"/bin/bash /tmp/restinterface/upload/nsf-odp-convert.sh %nsf-path% result.zip",
        "taskStatus":"EXECUTING"
    }  
```


Next, you need to asynchronously check the status every X seconds.  The operation should be fast, so we can keep X low - try X=5 for now

    # URL: /task/%taskid%
    curl "http://localhost:8095/task/%taskid%"

The from this response, you need to check 
- "taskStatus" - keep checking until this is "COMPLETED"  TODO:  other end states?
- "exitStatus" - TODO: do we need to check this?

```
    {
        "id":1,
        "startDate":"2022-06-27T22:31:13Z",
        "workingDir":"/tmp/restinterface/working",
        "command":"/bin/bash /tmp/restinterface/upload/nsf-odp-convert.sh /tmp/restinterface/upload/TestFieldOptions_1008_4.nsf result.zip",
        "output":"",
        "taskStatus":"COMPLETED",
        "exitStatus":0
    }
```

Once the action is confirmed successful, call this to download the resulting file.  You will need to save the response as a zip file:

    # URL:  /file/download
    # Parameter:  path - the remote path to the file, made by combining `%working-dir%` and %output-file%`
    curl --output local_result.zip "http://localhost:8095/file/download?path=%working-dir/%output-file%"

**Comment by rat-moonshine** [comment]
Updates:

- Crated initial UI/grid in `Settings > Vagrant`
- Initial VOs
- Add/edit/update of runtime instances

Now committing in a branch.

**Comment by JoelProminic** [comment]
@solaechea added the `/info` action to check the status and capabilities:

    % curl "http://localhost:8095/info"                                                            
    {
        "status":"ready",
        "capabilities":[
            "domino-api",
            "nsfconverter"
        ]
    }  

To determine if a server is available, Moonshine should check that "status" is "ready" (ignore case).  Treat other values, or a lack of response from the server, as "not available". 

"capabilities" should be checked based on the feature the user is trying to use.  For the Convert NSF action, check for the the "nsfconverter" value in "capabilities".

You will need to run "vagrant destroy" and "vagrant up" to get the update.  This is a bit annoying with the long Vagrantfile process, but I think it makes sense to do a clean update for this feature.

**Comment by rat-moonshine** [comment]
I confirm a pull, and then Vagrant destroy and up fix the problem. I can access http://localhost:8095/info correctly at the process end. Thank you @JoelProminic @solaechea .

**Comment by rat-moonshine** [comment]
Updates:

- New instance form validations 
- `/info` check and updating UI appropriately 
- Instance VOs save and retrieval
- In-progress Convert Database popup

**Comment by rat-moonshine** [comment]
I'm having some trouble uploading a file to local-server http://localhost:8095/file/upload. While the local server is accessible from browser, and also one of its agent (http://localhost:8095/info) from Moonshine, I see following errors while try uploading a file to the server:

```
[trace] httpStatusHandler: [HTTPStatusEvent type="httpStatus" bubbles=false cancelable=false eventPhase=2 status=500 redirected=false responseURL=null]
[trace] ioErrorHandler: [IOErrorEvent type="ioError" bubbles=false cancelable=false eventPhase=2 text="Error #2032: Stream Error. URL: http://localhost:8095/file/upload" errorID=2032]
```

I tested the upload with two separate APIs - `FileReference.upload(..)` and `URLLoader.load(..)` but both returns the same HTTPStatus and I/OError events.

@joshtynjala do you have any suggestion why this could be caused? The `FileReference.upload(..)` we are already using in Native and it seems to be work there.

**Comment by solaechea** [comment]
I see that there is an HTTP error 500, if that is returned from the rest-interface (http://localhost:8095/) it may be helpful to check the local-server logs in `/opt/rest-interface/log/output.log`

**Comment by joshtynjala** [comment]
> do you have any suggestion why this could be caused? The FileReference.upload(..) we are already using in Native and it seems to be working there. 
No idea.

> I wonder if it requires crossdomain policy file although it's an AIR app, so this might not require this.

AIR should never require a crossdomain.xml.

**Comment by rat-moonshine** [comment]
I'm having trouble or returns an error while try conversion request at http://localhost:8095/task:
 
```
: Failed to download file with exit code: 7
:   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
:                                  Dload  Upload   Total   Spent    Left  Speed
:   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (7) Failed connect to localhost:8080; Connection refused
```

**Comment by rat-moonshine** [comment]
Until now, I have completed till downloading `result.zip` file to user's Desktop (for now). Auto-unzip and opening into Moonshine-IDE is still pending.

However, I see some strange behaviour from http://localhost:8095/task calls. Following are the steps to describe the oddity:

1. Halt any running Vagrant instance
2. Run Vagrant instance from Moonshine
3. Go to `Others > Convert Domino Database`
4. Provide a .nsf file and Submit, in my case I tested with https://static.moonshine-ide.com/downloads/releases/dev/macos/branches/features/issue_1008_add_addtional_field_properties/TestFieldOptions_1008.nsf
5. Almost immediately all the intermediate process completes and a file being downloaded to user's Desktop as `result.zip` (**NOTE:** I see I need to print a more clear success download message to console)
6. Re-try the step from 3-4 again
7. This time http://localhost:8095/task never returns a `taskStatus=COMPLETE` but `taskStatus=EXECUTING` - even after waiting 4 minutes there were nothing change, as something stuck
8. Restart Moonshine - try 3-4 again, you'll see 7
9. Stop Vagrant and re-start
10. Re-try step 3-4
11. This time `result.zip` produced almost immediate again

My test observation is the conversion process properly returns only once after Vagrant instance re-starts. 

I have committed my changes and can be tested with macOS branch build from Bamboo.

**Comment by solaechea** [comment]
Awesome guys! 

Is it possible to use the Windows branch to test this? (I use Windows). 

It looks like it may be an issue in the _NSFConverterPortal_, I saw something similar but after a bigger number of conversions, I was able to do about 10 conversions and then the _NSFConverterPortal_ hanged. @rat-moonshine  you can see if the _NSFConverterPortal_ is up by opening the potal page in the local-server `http://localhost:8080/nsfConverter/main.xhtml` if that page doesn't load then it's highly probable that the _NSFConverterPortal_ is hanging. 

@JoelProminic If we add a parameter to the `POST /task` endpoint to override the default command timeout (which is 10 minutes), then we will get a response once that timeout is reached, the subprocess (curl) will be killed. Another option could be to set a timeout to curl in the [nsf-odp-convert.sh script](https://github.com/Moonshine-IDE/NSFConverterPortalProject/blob/ae8462a61e6024db955ecc83194ad9f3efdc7441/nsf-converter-portal/vagrant_support/nsf-odp-convert.sh)

**Comment by JoelProminic** [comment]
@JesseProminic, please do some tests with this when you have a chance.

I have links to the branch builds in the description.

Before you test, you will need to update NSFConverterPortalProject and recreate the Vagrant instance.
1. Make sure you have the new script:  `nsf-converter-portal/vagrant_support/nsf-odp-convert.sh`
2. In Moonshine right-click on nsf-converter-portal/Vagrant file, and choose Vagrant > Destroy
3. Then run Vagrant > Up on the same file.  It will probably take at least half an hour to run

Please test to make sure the process works for you.  Also, test if you can reproduce @rat-moonshine's error.

**Comment by JoelProminic** [comment]
@rat-moonshine, the problem you reported hasn't triggered for me.  I do see this behavior, but it seems to be something different (it may not be relevant anymore after you clean up the end of the process):
1. Start a conversion - this works fine
2. Close and reopen the prompt, and try again - this works fine
3. Retry without reopening the prompt - this hangs on the file upload.

We discussed the process a bit more today, and here is how we expect it to work:
1. Other > Convert Domino Dataqbase
2. Select a database ([examples](https://static.moonshine-ide.com/downloads/releases/dev/macos/branches/features/issue_1008_add_addtional_field_properties/))
3. Select a parent directory for the new project (new field)
4. Click Convert
5. The prompt closes, and the task progress is tracked in a sidebar, or in a new Vagrant tab in the footer (I'll explain this separately)
6. The user can cancel the task with a red stop button
7. Once the task finishes, Moonshine downloads the zip to a temporary location, and then unzips the project to the selected parent directory
8. Moonshine opens the project in the current workspace.

The Cancel button should close the prompt.  The current functionality will be moved to a "stop" button as described above.

**Comment by JoelProminic** [comment]
This is overkill for the moment, but in the long term we would like to have a tab to show the progress of the Vagrant tasks.  This should include at least:
- The Vagrant Instance that is used for the action
- A short description of the task (i.e. "Converting %database%.nsf" for this case)
- A button to kill the task

Some other information that could be displayed:
- An ID for each task?  I think that this is only needed if we could have multiple tasks that would otherwise look identical.   
- A progress tracker?  This could be based on the percentage of steps completed.  This is not an urgent priority
- Time the task has been running
- Split the interface and show the running Vagrant instances in the other half?

If you can easily add the task status to the taskbar at the bottom right (like we do for builds), this would be sufficent for.  If it would be easier to just work on the new Vagrant tab, I'm fine with that as well.

**Comment by JoelProminic** [comment]
I also did some testing on the Settings > Vagrant > Linked Instances interface.

Test | Result
-----|-------
New | Works  (should be "+"?)
Delete | TODO (should be "-"?)
Edit | Works (is the button redundant?)
Open | Not yet implemented
Status in prompt (server up) | Works, populated on focus change
Status in prompt (server down) | BUG:  triggers popup with "Error #2032".  No console prompt
Invalid Base URL - no protocol | BUG:  No error reported
Invalid Base URL - no port | BUG:  No error reported
Status=ready | 
Status!=ready | TODO:  I may need some help from @solaechea to setup a test for this
Not reachable | BUG:  State shows as blank
Capabilities | Populated as expected

Some updates for the Status:
- Change "State" to "Status" in the datagrid and prompt
- Remove the "Status" header in the prompt - this is now redundant
- If the `/info` action does not respond, show the state as "unreachable" with the error icon (#1045).  I chose "unreachable "instead of "offline" since it is not clear if the instance is offline or if the Base URL is invalid.  In the Status column, use a red icon.
- If the `/info` action returns "ready", show "ready" with a green checkbox or some other indication that this is a valid state.  In the Status column, use a green icon.
- For other statuses from /info, show the status with the error icon (#1045).  In the Status column, use a yellow icon, and show the actuall value in a tooltip.

And some tests for the NSF Converter combobox:

Test | Result
-----|-------
No Vagrant Instances | Blank
One valid instance | Works
Two valid instances | Works
Offline instance | Excluded
Online, but state is not "ready" | TODO
Missing nsf-converter Capability | TODO

**Comment by solaechea** [comment]
I was able to reproduce the issue described by Joel:
> 1. Start a conversion - this works fine
> 1. Close and reopen the prompt, and try again - this works fine
> 1. Retry without reopening the prompt - this hangs on the file upload.

But in the step 3 if you choose the NSF file again (without closing and reopening the "Convert Domino Database" prompt) the issue is not present.

I see an issue similar to what @rat-moonshine reported:
> 7. This time http://localhost:8095/task never returns a taskStatus=COMPLETE but taskStatus=EXECUTING - even after waiting 4 minutes there were nothing change, as something stuck

But in my case it's due to the NSFConverterPortal hanging. I was able to access `http://localhost:8080/nsfConverter/main.xhtml` when the task had about 3 minutes running (Not sure if the page was loaded from browser's cache), but after that I tried opening the portal page again and it didn't open (also hanged, keeps loading forever).

Moonshine-IDE keeps printing the Re-try conversion check until it reaches the 120 count (10 minutes)
![120 count](https://user-images.githubusercontent.com/1082804/177799761-f953b6a2-ea12-4920-8546-41e0b0ffdf9c.png)

After 10 minutes the Task reaches the timeout and an error is thrown in the rest-interface, the subprocess (curl) is never killed (I will check why) and appers to keep running "forever". The following screenshot is from the rest-interface logs:

![Captura de pantalla 2022-07-07 103814](https://user-images.githubusercontent.com/1082804/177801132-93eb8d9d-182e-4257-ba71-8224c49bf63e.png)

At that point, the `GET /task` request will return status=FAILED, the complete response is as follows (note that the error message is too verbose, so I took two screenshots of the same response):
![Captura de pantalla 2022-07-07 104825](https://user-images.githubusercontent.com/1082804/177803431-21a12bfa-df55-4a6c-9da2-a5b9ffa1a58f.png)
...
![Captura de pantalla 2022-07-07 104908](https://user-images.githubusercontent.com/1082804/177803465-06e37b23-499f-48fa-af2c-3ccfc24c6c1c.png)

You can find the complete response (JSON) in the attached zip file:
[get_task_JSON_error_nsf_portal_hang.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/9064873/get_task_JSON_error_nsf_portal_hang.zip)

About the error in the NSFConverterPortal, it looks like it may be an issue related to memory, the file download starts but it hangs at some point (Maybe because there are no more system resources for that application)?

**Comment by rat-moonshine** [comment]
> At that point, the GET /task request will return status=FAILED, the complete response is as follows (note that the error message is too verbose, so I took two screenshots of the same response):

I see `taskStatus=FAILED` wasn't managed at Moonshine side yet. So I can assume, unless user clicks on "Cancel" button in Moonshine popup/interface, Moonshine keeps calling `/task` agent in every 5 seconds which keeps it running forever.

With the new implementation as [suggested](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/979#issuecomment-1176618004) by @JoelProminic this shall be gracefully handle, user can terminate the check-process by clicking on the bottom-right taskbar button (which will be more obvious) and stops the `curl`/`/task` call.

**Comment by rat-moonshine** [comment]
I did database conversion process as [suggested](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/979#issuecomment-1176618004) by @JoelProminic with slight variation:

1. Other > Convert Domino Database
2. Select a database
3. Select a parent directory for the new project (new field)
4. Click Convert
5. [Variation] I intentionally do not close the popup immediately until the file uploading process completes. My reasoning was we should keep the prompt opens with all the fields populated until upload-success response returns at least - so an user can re-attempt to upload instead of re-opening the popup and set every fields again
6. Upload completes, the prompt closes, and the task progress is tracked in a sidebar, or in a new Vagrant tab in the footer
7. The user can cancel the task with a red stop button
8. Once the task finishes, Moonshine downloads the zip to a temporary location, and then unzips the project to the selected parent directory
9. Moonshine opens the project in the current workspace

While the above design mostly eliminates the problem I encountered to access `/task` - but not entirely. I can still reproduce the problem in following steps:

1. Convert a database and let's open the project in Moonshine
2. Delete the project, and re-attempt to convert
3. This time it also worked 
4. Delete the project again and re-attempt to convert
5. During 're-try' checks terminate the process by clicking on bottom-right Stop button
6. Re-attempt to convert 
7. This time I see it stuck again with endless 're-try' attempt on console unless I terminate the process
8. Re-start Moonshine, try to convert again
9. I see 7 

As it appears to me, unless a full-cycle completes at server side (uploading .nsf to downloading .zip), a new call of conversion fails somehow.

**Comment by rat-moonshine** [comment]
Earlier we had some problem unzipping the downloaded .zip file from Moonshine. I changed the unzip API to use one from MSDKI. 

I need someone to test on Windows who has everything setup already, to ensure that it properly unzips the downloaded project.

**Comment by rat-moonshine** [comment]
More updates:

- I have removed "Edit" labeled button, it is redundant. However, double-click to open/edit usually not a format we use in Moonshine
- Changed "Create" to "+" and "Remove" to "-" - however, using existing buttons styles a +/- sign looks faded
- "Open" button is now functional
- Icon with tooltip in "Status" column

**Comment by solaechea** [comment]
I just tested this in Windows 11. 
![about](https://user-images.githubusercontent.com/1082804/178001184-f94dddb0-075a-4d5f-b388-9d3d991fbeea.png)


### Test 1: Convert NSF with a parent directory distinct than Desktop
Selected Options:
![configuration](https://user-images.githubusercontent.com/1082804/177997267-5e86e58e-c1cb-44d5-afa5-70039b02ff1d.png)

Once I click submit the conversion starts, the popup is closed when the file uploading completes, I see the progress in the footer:
![Converting](https://user-images.githubusercontent.com/1082804/177997655-c06b07d7-4cd4-4f9c-a1fc-3f0b4d8517dc.png)

Once the conversion completes I see this error message:
![directory missing](https://user-images.githubusercontent.com/1082804/177997997-409edde4-fa52-4e68-af68-78446ffec1f5.png)

**I can see that the project folder was created (unzipped) in the selected Parent Directory to Download, but Moonshine is trying to open a folder in Desktop.**

If I Manually locate the unzipped folder and open it, the project opens correctly in Moonshine
![project opened](https://user-images.githubusercontent.com/1082804/177999400-b38ff2fd-bdf2-44fd-bf01-1e6d47f7e2dd.png)


### Test 2: Convert NSF with Desktop as parent directory
Selected Options:
![configuration 2 Desktop](https://user-images.githubusercontent.com/1082804/177999745-f0bcd46a-8eef-4491-8535-05eb421147ed.png)

I see a validation message, the folder (Parent Directory to Download/result) must be empty.
![destination should be empty](https://user-images.githubusercontent.com/1082804/178000450-bb4803ba-5d1e-4dae-840e-d8beff79024e.png)

Once I manually delete the folder at` .../Desktop/result`, I try again
This time the conversion completes and the project is automatically opened in Moonshine.
![project opened 2](https://user-images.githubusercontent.com/1082804/178001505-de5c3cee-8f27-4225-a290-aff85eede9f4.png)

### Test 3: Cancel NSF conversion by clicking the red square at the bottom
I was able to use the red square to stop the process
![red square](https://user-images.githubusercontent.com/1082804/178004179-f50410df-0766-4e1d-9aab-a279e999946d.png)

![stop process Moonshine](https://user-images.githubusercontent.com/1082804/178004124-61b0d771-3d04-4a82-92f1-c98807a005b9.png)

I would suggest adding a tooltip like "Stop process" to the stop button (red square).

**Comment by rat-moonshine** [comment]
Thank you @solaechea . I see my mistake where I [mistakenly](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/features/issue_979_vagrant_REST_interaction/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/vagrant/utils/ConvertDatabaseJob.as#L183) trying to open the project from Desktop as parent directory, and it was hard-coded. This was remaining from my earlier tests. 

I shall have this fixed.

**Comment by JoelProminic** [comment]
Currently the resulting project directory is "result".  I would like to update this to match the database file name (without `.nsf`), similar to how the `.veditorproj` is names.  So a database called `TestHideOptions.nsf` would result in a project named `TestHideOptions`.  We should watch for whitespace and other invalid project characters in the name.

During my tests, I got this error on an upload:

    : Error #2174
    : Error: Error #2174
    : 	at actionScripts.utils::FileUploader/startUpload()
    : 	at components.popup::ConvertDominoDatabasePopup/onSubmit()
    : 	at components.popup::ConvertDominoDatabasePopup/__btnSubmit_click()
    :
    : Click here to Report a Bug

The conversion process continued successfully anyway.  I'm not sure if this worked because I had previously uploaded the database or because the failed upload was a duplicate.

I couldn't reproduce this in later tests.  I tried:
- Multiple conversions without restarting Moonshine
- Uploading a database that had been uploaded before
- Double-clicking the Submit button.  This usually disappeared too quickly for me to get a reliable test.

**Comment by rat-moonshine** [comment]
> I see I [mistakenly](https://github.com/Moonshine-IDE/Moonshine-IDE/blob/features/issue_979_vagrant_REST_interaction/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/vagrant/utils/ConvertDatabaseJob.as#L183) trying to open the project from Desktop as parent directory, and it was hard-coded. This was remaining from my earlier tests.

This should be fixed now.

> Currently the resulting project directory is "result". I would like to update this to match the database file name (without .nsf), similar to how the .veditorproj is names. So a database called TestHideOptions.nsf would result in a project named TestHideOptions. We should watch for whitespace and other invalid project characters in the name.

Please, check.

> Error: Error #2174: Only one download, upload, load or save operation can be active at a time on each FileReference.

I added some additional handling to the file-uploader API. During my tests, I couldn't able to reproduce the issue, please give a test. 

Other updates:

- Vagrant instances' `status` now automatically checks while opening `Settings > Vagrant` and `Others > Convert Domino Database`
- Updating data-grid State column effectively

Please, check.

**Comment by JoelProminic** [comment]
Currently, I see Moonshine shows `Error #2032` in the Linked Vagrant Instance prompt and console when the server is unreachachable.  It also shows `Error #2032` on the console for each unreachable server when opening the Settings > Vagrant or Others > Convert Domino Database.

![image](https://user-images.githubusercontent.com/24532731/178323573-2b34b43c-f8a8-4c5e-bbaf-10187999d0ae.png)

However, I see this as an "expected" case.  The server may be unreachable if:
- The URL was entered incorrectly
- The REST Interface is not running for the instance
- The instance is currently down.  The user could shut down the instances when they are not in used to save system resources (this is also why I mentioned features like Open Project to facilitate restarting the instances).

So, I want to show "unreachable" for the "Status" (should be renamed from "State" in the Linked Vagrant Instance prompt) and no error on the console.

We could consider adding some debugging information in a tooltip.

**Comment by JoelProminic** [comment]
I did some more investigation for the file upload hang.  It looks like it is failing on [this line](https://github.com/Moonshine-IDE/NSFConverterPortalProject/blob/351751c6d8cad8f3b300a51ae15cf079b2baabe3/nsf-converter-portal/src/main/java/net/prominic/api/v1/DataService.java#L115)

I found [documentation](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#copy-java.io.InputStream-java.nio.file.Path-java.nio.file.CopyOption...-) for `Files.copy` here.  It reports a case where the method will hang:

> This method may block indefinitely reading from the input stream (or writing to the file). The behavior for the case that the input stream is asynchronously closed or the thread interrupted during the copy is highly input stream and file system provider specific and therefore not specified. 

I suspect we need to implement the call ourselves if we want to handle this case better.

**Comment by solaechea** [comment]
> I did some more investigation for the file upload hang. It looks like it is failing on [this line](https://github.com/Moonshine-IDE/NSFConverterPortalProject/blob/351751c6d8cad8f3b300a51ae15cf079b2baabe3/nsf-converter-portal/src/main/java/net/prominic/api/v1/DataService.java#L115)
> 
> I found [documentation](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#copy-java.io.InputStream-java.nio.file.Path-java.nio.file.CopyOption...-) for `Files.copy` here. It reports a case where the method will hang:
> 
> > This method may block indefinitely reading from the input stream (or writing to the file). The behavior for the case that the input stream is asynchronously closed or the thread interrupted during the copy is highly input stream and file system provider specific and therefore not specified.
> 
> I suspect we need to implement the call ourselves if we want to handle this case better.

@JoelProminic Thank you, I will take a look!

**Comment by JoelProminic** [comment]
Since I had some ideas for how to fix this, I added some code to implement a basic, low-level method for saving the uploaded file.  This method will fail on any Exceptions.

https://github.com/Moonshine-IDE/NSFConverterPortalProject/blob/96e36bdb93bb374f5a045d83bd8a9b2ea5ee4bb3/nsf-converter-portal/src/main/java/net/prominic/api/v1/DataService.java#L173-L204

This implementation is working for me.  If the bug triggers again, we should see more information in the log.

However, as part of testing this, I compared the original file to the file saved from NSFConverter, and I found that the checksum at least was different!

    # original
    [vagrant@localhost ~]$ ls -l /tmp/restinterface/upload/TestHideOptions_2022_04_25.nsf 
    -rw-rw-r--. 1 vagrant vagrant 458752 Jul 11 22:49 /tmp/restinterface/upload/TestHideOptions_2022_04_25.nsf
    [vagrant@localhost ~]$ md5sum /tmp/restinterface/upload/TestHideOptions_2022_04_25.nsf 
    8e88db90b57d171652e666b8bf7f208c  /tmp/restinterface/upload/TestHideOptions_2022_04_25.nsf
    # uploaded
    [vagrant@localhost ~]$ ls -l /tmp/TestHideOptions_2022_04_25.nsf
    -rw-rw-r--. 1 vagrant vagrant 458752 Jul 11 22:49 /tmp/TestHideOptions_2022_04_25.nsf
    [vagrant@localhost ~]$ md5sum /tmp/TestHideOptions_2022_04_25.nsf
    9142605bc1f4671e1b0da23b08cc8185  /tmp/TestHideOptions_2022_04_25.nsf

I wasted a lot of time with different tweaks for my code, and got different checksums each time.  I did a hex diff and found that there were a lot of small changes throughout the file.

Finally, I checked the original code again, and I found it had the same problem!  My theory is that the Notes API calls are making some small changes in the database by accessing it, so the file changes after the conversion runs.  For now, I committed my new code, but we can revisit this if we see other strange issues.

I also made some updates to improve the log configuration.  The log configuration can be updated [here](https://github.com/Moonshine-IDE/NSFConverterPortalProject/blob/96e36bdb93bb374f5a045d83bd8a9b2ea5ee4bb3/nsf-converter-portal/vagrant_support/payara_logging.properties).  The default log level for the `net.prominic` code is `INFO`, so we don't need to use `SEVERE` for all the log messages anymore.

**Comment by JoelProminic** [comment]
I haven't seen any more hangs since my updates yesterday.   Overall this is looking very good.  We can discuss tomorrow if we want to merge this.

@rat-moonshine made some changes after some discussion in the meeting today:
- Change alert button from "Ok" to "OK" when project exists - CONFIRMED
- Status wasn't updating properly when URL was updated in Linked Vagrant Instance - CONFIRMED FIXED
- Added icon and text for status in Linked Vagrant Instance - SEE BELOW

I haven't seen any more hangs since my updates yesterday.   Overall this is looking very good.  We can discuss tomorrow if we want to merge this.

I revisited the tests from https://github.com/Moonshine-IDE/Moonshine-IDE/issues/979#issuecomment-1176713825

#### Settings > Vagrant - there are a couple remaining tasks here.

Test | Result
-----|-------
New | Works
Delete | Works
Edit (Double-click | Works
Open | TODO:  Not yet implemented
Status in prompt (server up) | Works, populated on focus change and when opening for edit
Status in prompt (server down) | Works:  "unreachable"
Invalid Base URL - no protocol | TODO:  no validation
Invalid Base URL - no port |  TODO:  no validation
Status=ready | Works
Status!=ready | Works (yellow icon)
Not reachable | Works (red icon)
Capabilities | Populated as expected (including for status!="ready")

The statuses look like this.  The background for the green icon doesn't look good with the dark prompt background or highlighted datagrid rows.
![image](https://user-images.githubusercontent.com/24532731/178609328-6d9971de-05c4-47c6-9d2c-a7d2bd8f9f1b.png)
![image](https://user-images.githubusercontent.com/24532731/178609395-933bb8d0-fcab-4895-a07c-b53b2cbc052f.png)
![image](https://user-images.githubusercontent.com/24532731/178609535-11d0d08e-83cf-4cf0-ae58-ab6054f8bf60.png)

#### Combobox in Convert Domino Database prompt

Test | Result
-----|-------
No Vagrant Instances | Blank
One valid instance | Works
Two valid instances | Works
Offline instance | Excluded
Online, but status!="ready" | BUG:  This shows up in the combobox anyway if it has the nsfconverter capability
Missing nsfconverter Capability | Excluded

I uploaded my test for status!="ready" [here](https://static.moonshine-ide.com/downloads/releases/dev/macos/branches/features/issue_979_vagrant_REST_interaction/2022_07_12__Test-REST-Status-Starting.zip)

**Comment by rat-moonshine** [comment]
Updates:

- I have replaced the green-tick icon to something like this
![image](https://user-images.githubusercontent.com/24370870/178673018-a546ca47-7a29-47b1-b106-82ca129f547c.png)
- There was an URL validation already, but I noticed that it was not marking an invalid URL on UI, nor it was preventing the form from submit when it's invalid - this should fix now
- Dropdown now also checks and includes instance that has `state == ready`

Please, check.

**Comment by JoelProminic** [comment]
These updates look good:
- icons - all of the icons are clear now.
- Applications in a status!=ready state are excluded from the Convert Domino Database combobox
- Open works if a path was specified.  It is disabled if no path is set.   If the project doesn't exist, Moonshine will create the directory and try to open it as a generic project.  This should probably show an error instead, but it should only happen if the user moves or removes the corresponding project.


I see the URL validation looks like this:
![image](https://user-images.githubusercontent.com/24532731/179068018-5d3cdba8-c56c-4306-9418-e3f40f93549a.png)

I think this is too subtle.  In order to see the error, I need to check the tooltip, and if I miss the red outline, there is no indication on Save for why the action failed.  Perhaps we could display the error message in the Status in this case?

We need a bit more information in the error message:

> Invalid URL format.  Expected format is http(s)://address:port

I also noticed that port is required to be at least 4 characters long.  I think this is a reasonable restriction - I don't expect we would use less characters.

**Comment by rat-moonshine** [comment]
Updates:

- "Open" now prints error message to console if the path is invalid instead of trying opening it
- URL invalid message is more clear now:
![image](https://user-images.githubusercontent.com/24370870/179164229-72dcf6dd-2670-4367-9212-fa2591f1cf77.png)

Please, check.

**Comment by rat-moonshine** [comment]
Updates:

- "Open" now prints error message to console if the path is invalid instead of trying opening it
- URL invalid message is more clear now:
![image](https://user-images.githubusercontent.com/24370870/179164752-57642fb8-95dd-4bab-8d91-7bd22aa9ceb0.png)

Please, check.

**Comment by JoelProminic** [comment]
The updates look good.  I think this is ready for a merge.

**Comment by rat-moonshine** [comment]
This is now merged into `master`.

**Comment by JoelProminic** [comment]
I briefly tested the merged feature on macOS, and it seems fine.   @feather812002, you may find this feature convenient for testing NSFConverter after you merge master into your current development branch.   If necessary, I can walk you through the setup in the Sametime meeting.

I also did some tests with the Windows build, and this seems to be working properly as well.   

Some notes about remote connections:  I tested this from a Vagrant Windows instance, and in order to reach the NSFConverter instance on the same machine I needed to use the internal IP.  However, the current `nsf-conveter-portal/Vagrantfile` limits the connections to come from 127.0.0.1 (using `host_ip`).  If we ant to support an internal IP, we need to add an entry for the internal IP.

  config.vm.network "forwarded_port", guest: REST_INTERFACE_HTTP_GUEST_PORT, host: 8195, host_ip: "192.168.x.x"

Alternatively, we can leave `host_ip` blank to allow access from any IP.  I'm concerned about the security for this, but it could be protected by the firewall instead.  See [this page](https://www.vagrantup.com/docs/networking/forwarded_ports) for more information.

**Comment by JoelProminic** [comment]
The core functionality is working for this, so I'm closing the issue.
